### PR TITLE
[Federation] Make arguments to scheduling type adapter methods generic

### PR DESCRIPTION
**Pull request by irfanurrehman**
This is in the process of trying to rebase https://github.com/kubernetes/kubernetes/pull/45993 on latest.
cc @marun @perotinus 
@kubernetes/sig-federation-misc 
Hoping I get some attention to this and later PRs soon.

Associated issue https://github.com/kubernetes/kubernetes/issues/49181

**Release note**:

```NONE
```

### CODE REVIEW

**Comment by marun** [comment]
I think you'll want to base your work on #46260, since it makes the scheduling logic common between replicaset and deployment.

**Comment by irfanurrehman** [comment]
> I think you'll want to base your work on #46260, since it makes the scheduling logic common between replicaset and deployment.

@marun, I see that the deployment controller PR is also merged; I have rebased this on latest thus. This however is only making some arguments, which were quite specific to rs and deployment,  within the scheduling adapter methods generic. 

Additionally, If you remember our pre 1.7 code freeze discussions, I still find this mechanism of providing a facility for controller implementation to hook into sync controller reconcile, a little non generic and a little flawed as in my comments earlier; for example https://github.com/kubernetes/kubernetes/pull/45563#discussion_r117958806 and https://github.com/kubernetes/kubernetes/pull/45563#discussion_r118777573. I still do not get the reason of passing controller specific info (named schedulinginfo) back and forth into the reconcile of sync controller. Wouldn't it be easier for the implementer of the new controllers not to worry about creating/formatting/passing this information to adhere to this scheduling adapter type. I think it would be easier for the implementer to unit test the code also, without the additional complexity.
I also tried to break my head around to consider this piece of code as for scheduling type alone (say rs and deployment) and add another interface which just implements one hook into the sync controller reconcile, but another types attached to sync controller will need additional if - else type checks in the sync controllers reconcile, which sounds a little lame. I am thinking from the point of view of new controllers which need to be coded on top of sync (hpa, service, statefulset, ...) and could still think of only one interface method solution (as in this PR https://github.com/kubernetes/kubernetes/pull/45497) to be most simple and most generic. All these controllers (hpa, service, statefulset, ...) need to do more stuff then distribute replicas. Additionally the code in https://github.com/kubernetes/kubernetes/pull/45497 is not the best, might lack tests and can certainly be better organised, but the intent in it is clear.
On the other hand, I think I can probably align https://github.com/kubernetes/kubernetes/pull/45993 to the current code in master using SchedulingAdapter, but the statefulset controller will neither align to the interface method names completely, nor will I be able to align the code logically into the SchedulingAdapter interface methods.
I am not really sure if https://github.com/kubernetes/kubernetes/pull/45497 is the best solution for aligning all types needing more stuff then sync; I can think only so much with my limited undertanding.
I am all willing to consider alternatives or suggestions and also willing to spend time working on anything that might be needed to have one common solution for all controllers which need to do more then normal sync, before rebasing my work on top. Please advice!

cc @perotinus, @quinton-hoole, @csbell

**Comment by perotinus** [comment]
@irfanurrehman At this point, I believe that you have the most context of this problem, given the merged approach for replica sets and deployments and your in-flight work. I'm not attached to the way that the scheduling logic currently works, and if from your perspective you can't make the current approach work for your needs, and believe that you can make your approach work here without adding unnecessary complexity, then I think that's probably the best thing going forward.

I would question, though, whether the HPA and stateful set controllers are necessarily "scheduling controllers." In other words, is there a kernel of scheduling that the replicaset/jobs/deployment controllers share, while the HPA and stateful set controllers have additional logic that can be represented in a different interface; or, another way, are the HPA and statefeul set controllers "scheduling+X" controllers?  I'm not sure that this is the best way to go about figuring this out or representing it in the code, but there may be enough conceptual differences between these controllers conceptually that trying to fit them all into the same interface will make the code too generic.

**Comment by irfanurrehman** [comment]
> @irfanurrehman At this point, I believe that you have the most context of this problem, given the merged approach for replica sets and deployments and your in-flight work. I'm not attached to the way that the scheduling logic currently works, and if from your perspective you can't make the current approach work for your needs, and believe that you can make your approach work here without adding unnecessary complexity, then I think that's probably the best thing going forward.
> 
> I would question, though, whether the HPA and stateful set controllers are necessarily "scheduling controllers." In other words, is there a kernel of scheduling that the replicaset/jobs/deployment controllers share, while the HPA and stateful set controllers have additional logic that can be represented in a different interface; or, another way, are the HPA and statefeul set controllers "scheduling+X" controllers? I'm not sure that this is the best way to go about figuring this out or representing it in the code, but there may be enough conceptual differences between these controllers conceptually that trying to fit them all into the same interface will make the code too generic.

Thanks @perotinus for having a look at this. I completely understand what you say here. I also quite like the code that u have updated to make the rs and deployment controllers into one single scheduling add on controller. Its indeed very nicely done for the two. In the meanwhile as mentioned in my earlier comment, hpa stuff can with some effort be adapted to what you wrote. I am almost done with that adaption. It makes some tweaks additional to this PR which I will submit on the hpa PR with the hpa controller. I am adapting the unit tests for the same. Will submit the whole PR soon. Request you to spare some time to have a look at that later.
Only if the need be later; I would separately submit updates hpa controller to make the sync plugin generic enough to be usable by further controllers.
This current PR is usable in itself, so it will be great, if u can review this as is (It changes only some types and names). Thanks again!

**Comment by perotinus** [comment]
These changes LGTM code wise. I don't have a deep knowledge of the HPA controller, but I can believe that it's different enough that it requires a different interface.

@marun, do you have any concerns here? I think, at this point, this approach seems like a reasonable step, and we can always look back and make things cleaner if and when it comes to that.

**Comment by irfanurrehman** [comment]
@marun @perotinus, if you have no other comments, can u please lgtm this.
@kubernetes/sig-federation-pr-reviews can any of the approvers please approve this.

File: federation/pkg/federation-controller/sync/controller.go [hunk 1]
```diff
@@ -490,10 +490,13 @@ func syncToClusters(clustersAccessor clustersAccessorFunc, operationsAccessor op
 		if !ok {
 			glog.Fatalf("Adapter for kind %q does not properly implement SchedulingAdapter.", kind)
 		}
-		err = schedulingAdapter.UpdateFederatedStatus(obj, schedulingInfo.Status)
-		if err != nil {
-			runtime.HandleError(fmt.Errorf("adapter.UpdateFinished() failed on adapter for %s %q: %v", kind, key, err))
-			return statusError
+		if schedulingInfo != nil {
```
Comment (csbell) [inline federation/pkg/federation-controller/sync/controller.go]: Seems odd to check for a non-nil schedulingInfo given that we have an error check above. Is a nil-defined schedulingInfo a valid state? If so, please document.

Comment (irfanurrehman) [inline federation/pkg/federation-controller/sync/controller.go]: I did put this check while changing the params to/from schedule adapter to interfaces which passes pointer.
You are right, this check is not at all needed as none of the current implementations actually pass a null. Removed it!

**Comment by irfanurrehman** [comment]
@csbell in case you miss the reply to the comment (given how github review works). I have updated this as per your comment!

**Comment by irfanurrehman** [comment]
@csbell I guess the do-not-merge label will need to be removed. Thanks!

**Comment by perotinus** [comment]
@irfanurrehman Do you have an issue you can associate this with? That's what's preventing the approval.

**Comment by irfanurrehman** [comment]
> @irfanurrehman Do you have an issue you can associate this with? That's what's preventing the approval.

Thanks @perotinus, I did not have one, but created and referenced now!

**Pull request merged**